storage/factory/pruningStorerFactory.go

```diff
@@ -0,0 +1,537 @@
+		PersisterFactory:      NewPersisterFactory(psf.generalConfig.BootstrapStorage.DB),
+		BloomFilterConf:       GetBloomFromConfig(psf.generalConfig.BootstrapStorage.Bloom),
+		NumOfEpochsToKeep:     numOfEpochsToKeep,
+		NumOfActivePersisters: numOfActivePersisters,
+		Notifier:              psf.epochStartNotifier,
+	}
+	bootstrapUnit, err = pruning.NewPruningStorer(bootstrapUnitArgs)
+	if err != nil {
+		return nil, err
+	}
+	successfullyCreatedStorers = append(successfullyCreatedStorers, bootstrapUnit)
+
+	store := dataRetriever.NewChainStorer()
+	store.AddStorer(dataRetriever.TransactionUnit, txUnit)
+	store.AddStorer(dataRetriever.MiniBlockUnit, miniBlockUnit)
+	store.AddStorer(dataRetriever.PeerChangesUnit, peerBlockUnit)
+	store.AddStorer(dataRetriever.BlockHeaderUnit, headerUnit)
+	store.AddStorer(dataRetriever.MetaBlockUnit, metachainHeaderUnit)
+	store.AddStorer(dataRetriever.UnsignedTransactionUnit, unsignedTxUnit)
+	store.AddStorer(dataRetriever.RewardTransactionUnit, rewardTxUnit)
+	store.AddStorer(dataRetriever.MetaHdrNonceHashDataUnit, metaHdrHashNonceUnit)
+	hdrNonceHashDataUnit := dataRetriever.ShardHdrNonceHashDataUnit + dataRetriever.UnitType(psf.shardCoordinator.SelfId())
+	store.AddStorer(hdrNonceHashDataUnit, shardHdrHashNonceUnit)
+	store.AddStorer(dataRetriever.HeartbeatUnit, heartbeatStorageUnit)
+	store.AddStorer(dataRetriever.BootstrapUnit, bootstrapUnit)
+
+	return store, err
+}
+
+// CreateForMeta will return the storage service which contains all storers needed for metachain
+func (psf *StorageServiceFactory) CreateForMeta() (dataRetriever.StorageService, error) {
+	var shardDataUnit *pruning.PruningStorer
+	var metaBlockUnit *pruning.PruningStorer
+	var headerUnit *pruning.PruningStorer
+	var txUnit *pruning.PruningStorer
+	var peerDataUnit *pruning.PruningStorer
+	var metaHdrHashNonceUnit *pruning.PruningStorer
+	var miniBlockUnit *pruning.PruningStorer
+	var unsignedTxUnit *pruning.PruningStorer
+	var miniBlockHeadersUnit *pruning.PruningStorer
+	var shardHdrHashNonceUnits []*pruning.PruningStorer
+	var bootstrapUnit *pruning.PruningStorer
+	var err error
+
+	successfullyCreatedStorers := make([]storage.Storer, 0)
+
+	defer func() {
+		// cleanup
+		if err != nil {
+			log.Error("create meta store", "error", err.Error())
+			for _, storer := range successfullyCreatedStorers {
+				_ = storer.DestroyUnit()
+			}
+			if shardHdrHashNonceUnits != nil {
+				for i := uint32(0); i < psf.shardCoordinator.NumberOfShards(); i++ {
+					if shardHdrHashNonceUnits[i] != nil {
+						_ = shardHdrHashNonceUnits[i].DestroyUnit()
+					}
+				}
+			}
+		}
+	}()
+
+	fullArchiveMode := psf.generalConfig.StoragePruning.FullArchive
+	numOfEpochsToKeep := uint32(psf.generalConfig.StoragePruning.NumEpochsToKeep)
+	numOfActivePersisters := uint32(psf.generalConfig.StoragePruning.NumActivePersisters)
+
+	metaBlockUnitArgs := &pruning.PruningStorerArgs{
```

Comment:
can you extract the pruning storer arguments creation in a method? as I see you only need as parameter the identifier and the storer you want to create arguments for, as the other params you have in the receiver.

this would remove a lot of duplicated initialization code.

Reply:
done

Reply:
looks better, can be refactored more but it can go in as it is.